[GStreamer][MSE] Avoid adding redundant tracks to MediaPlayer

vivienne-w opened the pull request:
#### efcd4ab82387046eaa2a8534c34e085d32c41a25
<pre>
[GStreamer][MSE] Avoid adding redundant tracks to MediaPlayer
<a href="https://bugs.webkit.org/show_bug.cgi?id=288974">https://bugs.webkit.org/show_bug.cgi?id=288974</a>

Reviewed by Alicia Boya Garcia.

In MSE, WebKit already adds tracks to the HTMLMediaElement and the SourceBuffer in
SourceBuffer::sourceBufferPrivateDidReceiveInitializationSegment, so we need to
avoid adding each Track again in updateTracks.

* LayoutTests/media/media-source/media-source-vttsimple-expected.txt:
* LayoutTests/media/media-source/media-source-vttsimple.html: Explicitly test for the number of tracks.
* LayoutTests/media/media-source/media-source-webm-expected.txt:
* LayoutTests/media/media-source/media-source-webm.html: Ditto.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::updateTracks): Don&apos;t add Tracks to MediaPlayer if in MSE.

Canonical link: <a href="https://commits.webkit.org/291559@main">https://commits.webkit.org/291559@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ade8c271ad9bf6b51234946eb6dd4ba5448ab02e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93290 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12848 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2573 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98289 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43816 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95340 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13135 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21302 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71293 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28687 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96292 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9856 "Found 1 new test failure: fast/forms/ios/focus-input-via-button-no-scaling.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84385 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51626 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9548 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/2017 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43130 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/79834 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2031 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100320 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20342 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14888 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80313 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20594 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80305 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79628 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19785 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24173 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1506 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/13459 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20326 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25503 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20013 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23473 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21754 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ntrrgc commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -1739,7 +1739,7 @@ void MediaPlayerPrivateGStreamer::updateTracks([[maybe_unused]] const GRefPtr<Gs
+            if (player && !useMediaSource)                              \
                 player->add##Type##Track(track);                        \
             m_##type##Tracks.add(streamId, WTFMove(track));             \
         }                                                               \
```
I am questioning whether we should create and store these `*TrackPrivateGStreamer` instances for the MSE case.

Think about it: we're fixing the duplicate track bug by not reporting these to the multiplatform player code, but we're still creating and storing them in `m_audioTracks`, `m_videoTracks` and we would be storing them in `m_textTracks` as well if not for this:

```
        else if (type & GST_STREAM_TYPE_TEXT && !useMediaSource)
            CREATE_OR_SELECT_TRACK(text, Text);
```

But with what we have figured out so far, do we even need any of the code in `MediaPlayerPrivateGStreamer::updateTracks()` at all for MSE?

All the usages I see look like stuff the multiplatform code could be handling already in the case of MSE, and the only side effect of `updateTracks()` that looks obviously relevant is the code in `scopeExit`: updating `m_hasAudio` and `m_hasVideo` and calling `characteristicChanged()`, `sizeChanged()` and `mediaEngineUpdated()` if needed, but we could do that without creating any "duplicate" track objects.

vivienne-w replied:
Yeah, makes sense. I've added a FIXME for now, and I'll try to get back to this soon-ish

ntrrgc reviewed (approved):
Good catch! LGTM.

Please look at what I mentioned in the comment. It doesn't have to be done in this patch, but if not, it would be a good idea to leave a FIXME comment somewhere so that it doesn't get lost.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ade8c271ad9bf6b51234946eb6dd4ba5448ab02e. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/41777

vivienne-w committed:
[GStreamer][MSE] Avoid adding redundant tracks to MediaPlayer
https://bugs.webkit.org/show_bug.cgi?id=288974

Reviewed by Alicia Boya Garcia.

In MSE, WebKit already adds tracks to the HTMLMediaElement and the SourceBuffer in
SourceBuffer::sourceBufferPrivateDidReceiveInitializationSegment, so we need to
avoid adding each Track again in updateTracks.

* LayoutTests/media/media-source/media-source-vttsimple-expected.txt:
* LayoutTests/media/media-source/media-source-vttsimple.html: Explicitly test for the number of tracks.
* LayoutTests/media/media-source/media-source-webm-expected.txt:
* LayoutTests/media/media-source/media-source-webm.html: Ditto.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::updateTracks): Don't add Tracks to MediaPlayer if in MSE.

Canonical link: https://commits.webkit.org/291559@main

webkit-commit-queue commented:
Committed 291559@main (efcd4ab82387): <https://commits.webkit.org/291559@main>

Reviewed commits have been landed. Closing PR #41777 and removing active labels.

aperezdc commented:
Backported into `webkitglib/2.48` as commit e612704d066c9911a5673dc32954bb6973da59c9
